server/plugin_release.go

```diff
@@ -29,199 +35,406 @@ import (
+		return fmt.Errorf("failed to decode public key err=%w", err)
+	}
+
+	keyring, err := openpgp.ReadKeyRing(block.Body)
+	if err != nil {
+		return fmt.Errorf("can't read public key err=%w", err)
+	}
+
+	for _, pluginFilePath := range pluginFilePaths {
+		signedFile, err := os.Open(pluginFilePath)
+		if err != nil {
+			return fmt.Errorf("cannot read signed file err=%w", err)
+		}
+		defer signedFile.Close()
+
+		// Assume signature is always <filepath>.sig
+		signatureFile, err := os.Open(fmt.Sprintf("%s.sig", pluginFilePath))
+		if err != nil {
+			return fmt.Errorf("cannot read signature file err=%w", err)
+		}
+		defer signatureFile.Close()
+
+		if _, err = openpgp.CheckDetachedSignature(keyring, signedFile, signatureFile); err != nil {
+			return fmt.Errorf("error while checking the signature err=%w", err)
+		}
 	}
 
-	return releaseAsset.GetBrowserDownloadURL(), nil
+	LogInfo("Signatures verified for %+v", pluginFilePaths)
+	return nil
 }
 
-func uploadSignedArtifcatsToGithub(ctx context.Context, githubClient *github.Client, repo string, releaseID int64, tempFolder string, fileToUpload []string) error {
-	LogInfo("Uploading signed assets to Github release")
+// createPlatformPlugins splits plugin tar into platform specific plugin tars.
+// Returns paths to platform plugin tars if successful, or an error otherwise.
+func createPlatformPlugins(repositoryName, tag, pluginFilePath, pluginFolder string) ([]string, error) {
+	var result []string
+	if err := hasAllPlatformBinaries(pluginFilePath); err != nil {
+		return nil, err
+	}
 
-	for _, file := range fileToUpload {
-		opts := &github.UploadOptions{
-			Name: file,
+	platformFileExclusion := map[string][]string{
+		"osx-amd64":     {"windows", "linux"},
+		"windows-amd64": {"darwin", "linux"},
+		"linux-amd64":   {"darwin", "windows"},
+	}
+
+	platformFileInclusion := map[string]string{
+		"osx-amd64":     "plugin-darwin-amd64",
+		"windows-amd64": "plugin-windows-amd64.exe",
+		"linux-amd64":   "plugin-linux-amd64",
+	}
+
+	for platform, excludeOtherPlatforms := range platformFileExclusion {
+		deleteWildcards := ""
+		for _, deleteWildcard := range excludeOtherPlatforms {
+			deleteWildcards += fmt.Sprintf(`--delete "*%s*" `, deleteWildcard)
 		}
 
-		filePath := fmt.Sprintf("/%s/%s", tempFolder, file)
-		fileHandler, err := os.Open(filePath)
-		defer fileHandler.Close()
-		if err != nil {
-			LogError("Error opening the file. err=" + err.Error())
-			return err
+		// Couldn't achieve gzip level compressions with golang archive api, using shell cmds instead.
+		platformTar := filepath.Join(pluginFolder, fmt.Sprintf("%v-%v-%v.tar.gz", repositoryName, tag, platform))
+		tarCmd := "tar"
+		if runtime.GOOS == "darwin" {
+			tarCmd = "gtar"
 		}
-		_, _, err = client.Repositories.UploadReleaseAsset(ctx, Cfg.GithubOrg, repo, releaseID, opts, fileHandler)
+		cmd := fmt.Sprintf(`cat %s | gunzip | %s --wildcards %s | gzip > %s`, pluginFilePath, tarCmd, deleteWildcards, platformTar)
```

Comment:
Ahh good eye! This could be tricky 🤔. `pluginFilePath` and `platformTar` may contain some funky characters, best bet to sanitize? What do you think @jupenur?

Totally agreed that we should avoid `exec` if we can, my first attempt was to avoid it, but then i wasn't getting low enough compressions, which is why i had to switch to this.

Reply:
Yeah this approach definitely calls for sanitization, but doing that for shell commands can be surprisingly difficult. If we can't avoid `exec`, could we at least avoid `bash -c` and all the piping so that individual commands and args could be passed separately and not as one long string?

Reply:
After wrestling with golang for half a day, finally got it working with golang piping api.